keps/sig-scheduling/3022-min-domains-in-pod-topology-spread/README.md

```diff
@@ -0,0 +1,535 @@
+
+### Goals
+
+- Users can specify `minDomains` to limit the number of domains.
+- Users can use it as a mandatory requirement with `WhenUnsatisfiable=DoNotSchedule`.
+- Users can use it as a best-efforts manner with `WhenUnsatisfiable=ScheduleAnyway`.
+
+### Non-Goals
+
+- Add new field to limit the maximum number of topology domains.
+
+## Proposal
+
+### User Story
+
+I am using cluster autoscaler and I want to force spreading a deployment over at least 5 Nodes.
+
+## Design Details
+
+Users can define a minimum number of domains with `minDomains` parameter.
+
+Pod Topology Spread has the semantics of "global minimum", which means the minimum number of pods that match the label selector in a topology domain.
+
+When the number of domains with matching topology keys is less than `minDomains`,
+Pod Topology Spread treats "global minimum" as 0; other wise, "gloval minimum"
+is equal to the minimum number of matching pods on a domain.
+
+As a result,
+
+- when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same or more number of matching Pods as `maxSkew`.
+- when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes on the domains that have the less number of matching Pods as `maxSkew`.
+
+`minDomains` is an optional parameter and the default value is 0.
+
+### API
+
+New optional parameter called `MinDomains` is introduced to `PodSpec.TopologySpreadConstraint[*]`.
+
+```go
+type TopologySpreadConstraint struct {
+......
+  // When the number of domains with matching topology keys is less than `minDomains`,
+  // Pod Topology Spread treats "global minimum" as 0.
+  // As a result,
+  // - when `whenUnsatisfiable` equals `DoNotSchedule`,
+  //   scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same number of matching Pods as `maxSkew`.
+  // - when `whenUnsatisfiable` equals `ScheduleAnyway`,
+  //   scheduler prefers Nodes on the domains that don't have the same number of matching Pods as `maxSkew`.
+  // Default value is 0.
+  // +optional
+  MinDomains *int32
+}
+```
+
+### Implementation details
+
+**Filtering phase**
+
+In Filter of Pod Topology Spread, filtering criteria is
+
+```
+('existing matching num' + 'if self-match (1 or 0)' - 'global min matching num') <= 'maxSkew'
+```
+
+- `existing matching num` denotes the number of current existing matching Pods on the domain.
+- `if self-match` denotes if the labels of Pod is match with selector of the constraint.
+- `global min matching num` denotes the minumun number of matching Pods.
+
+For `whenUnsatisfiable: DoNotSchedule`, Pod Topology Spread treats `global min matching num` as 0
+when the number of domains with matching topology keys is less than `minDomains`.
+
+In `preFilterState`, Pod Topology Spread has `Constraints` field which type is slice of internal `topologySpreadConstraint` struct.
```

Comment:
I think the important question that the "implementation details" needs to answer is how to calculate the number of matching domains. If this is already calculated, say where it is already calculated and that we can compare that to `minDomains` to set `global min = 0`. Then the formula above applies directly.

Reply:
There was a slight misunderstanding and I didn't explain it well enough. Sorry.
I've added the explanation.

Reply:
Keep in mind that readers of a KEP are not necessarily familiar with low level details in kube-scheduler (like production readiness reviewer). So we should try to keep the explanations as light as possible.

You can just say something like this:

We can calculate the number of matching domains in `PreFilter`, along with the calculation of `TpPairToMatchNum` (link here). This extra calculation doesn't increase the complexity of the preFilter logic.

Reply:
Sure, I made the description for Filter simpler as suggestion.